iphone/Maps/UI/Downloader/MWMMapDownloaderViewController.mm

```diff
@@ -124,40 +138,40 @@ - (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar
 
 - (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
 {
-  if (searchText.length == 0)
-  {
-    self.dataSource = self.defaultDataSource;
-    [self reloadTable];
-  }
-  else
-  {
-    NSString * primaryLanguage = self.searchBar.textInputMode.primaryLanguage;
-    if (primaryLanguage)
-      m_searchParams.m_inputLocale = primaryLanguage.UTF8String;
-
-    m_searchParams.m_query = searchText.precomposedStringWithCompatibilityMapping.UTF8String;
-    GetFramework().SearchInDownloader(m_searchParams);
-  }
+  NSString * primaryLanguage = self.searchBar.textInputMode.primaryLanguage;
+  if (primaryLanguage)
+    m_searchParams.m_inputLocale = primaryLanguage.UTF8String;
+
+  m_searchParams.m_query = searchText.precomposedStringWithCompatibilityMapping.UTF8String;
+  [self updateSearchCallback];
+  GetFramework().SearchInDownloader(m_searchParams);
 }
 
 #pragma mark - UIBarPositioningDelegate
 
 - (UIBarPosition)positionForBar:(id<UIBarPositioning>)bar { return UIBarPositionTopAttached; }
 #pragma mark - Search
 
-- (void)setupSearchParams
+- (void)updateSearchCallback
 {
   __weak auto weakSelf = self;
-  m_searchParams.m_onResults = ^(DownloaderSearchResults const & results) {
+  NSTimeInterval const timestamp = [NSDate date].timeIntervalSince1970;
+  self.lastSearchTimestamp = timestamp;
+  m_searchParams.m_onResults = [weakSelf, timestamp](DownloaderSearchResults const & results) {
     __strong auto self = weakSelf;
-    if (!self || results.m_endMarker)
+    if (!self || timestamp != self.lastSearchTimestamp)
       return;
-    self.searchDataSource =
-        [[MWMMapDownloaderSearchDataSource alloc] initWithSearchResults:results delegate:self];
-    dispatch_async(dispatch_get_main_queue(), ^{
+    if (results.m_query.empty())
+    {
+      self.dataSource = self.defaultDataSource;
+    }
+    else
+    {
+      self.searchDataSource =
```

Comment:
We have 2 datasources in this controller:
* searchDataSource - (search results)
* defaultDataSource - (list of available or downloaded maps at current level)

And we have dataSource - the pointer to the active datasource.
Controller uses dataSource to display table.